Adds thread runner, benchmark and stats components.

BrianSong opened the pull request.

BrianSong committed:
Adds thread runner, benchmark and stats components.

BrianSong committed:
Update thread_runner.cc

BrianSong committed:
Fixed BUILD.

BrianSong committed:
Merge branch 'develop' of https://github.com/BrianSong/ml-metadata into develop

hughmiao commented on ml_metadata/tools/mlmd_bench/benchmark.h:
```diff
@@ -0,0 +1,45 @@
+class Benchmark {
+ public:
+  Benchmark(const MLMDBenchConfig& mlmd_bench_config);
+  ~Benchmark();
```
nit, use = default, the vector and unique-ptr deconstructor will be simply called

(see other .h for examples when having a trivial deconstructor)

BrianSong replied:
Fixed! Thanks.

hughmiao commented on ml_metadata/tools/mlmd_bench/benchmark.h:
```diff
@@ -0,0 +1,45 @@
+  // Takes `workload_config` and generates corresponding executable workloads.
+  void CreateWorkload(const WorkloadConfig& workload_config);
+
+  std::vector<std::pair<std::unique_ptr<WorkloadBase>, int64>>& workloads();
```
often return unique_ptr pointer to caller means transfer ownership, use vector of WorkloadBase*  instead. 
or better, use a size(), and workloads(i) to get each workload separately, it is not necessary to expose the implementation details of the workloads. wdyt?

BrianSong replied:
Looks good to me! Fixed :)

hughmiao commented on ml_metadata/tools/mlmd_bench/benchmark.h:
```diff
@@ -0,0 +1,45 @@
+
+ private:
+  // A list of executable workloads and their according number of operations.
+  std::vector<std::pair<std::unique_ptr<WorkloadBase>, int64>> workloads_;
```
qq, why add the int64 here?, workload should have a num_operations member to get it?

BrianSong replied:
Yes, but as what we designed before, the num_operation() member function in Workload class will return work_items_.size(). Here, we have not setup the workload yet, so num_operation() will not work here. Or do you think we should add a additional getter function to get num_operations_? wdyt?

hughmiao commented on ml_metadata/tools/mlmd_bench/benchmark.cc:
```diff
@@ -0,0 +1,58 @@
+  }
+}
+
+Benchmark::~Benchmark() {
+  for (auto& workload : workloads_) {
+    workload.first.reset();
+  }
+}
```
see comments in .h

BrianSong replied:
Fixed!

hughmiao commented on ml_metadata/tools/mlmd_bench/benchmark.h:
```diff
@@ -0,0 +1,45 @@
+  Benchmark(const MLMDBenchConfig& mlmd_bench_config);
+  ~Benchmark();
+
+  // Takes `workload_config` and generates corresponding executable workloads.
+  void CreateWorkload(const WorkloadConfig& workload_config);
```
should this be a internal anonymous namespace method?
also consider to use a template CreateWorkload(const T& config, ...), where the T is the one of those WorkloadConfigs, use a list of specialized CreateWorkload<FillTypeWorkloadConfig) to implement them separately. wdyt?

BrianSong replied:
Yes, my bad. It should be a internal anonymous namespace method. However, as for the template method, will it be too much difference compared to the current method? Since every workload has different constructor to pass in which means we have to create many branches as well? wdyt?

hughmiao commented on ml_metadata/tools/mlmd_bench/stats.h:
```diff
@@ -20,13 +20,55 @@ limitations under the License.
+  // per seconds for the current workload.
+  void Report(const std::string& specification);
+
+  absl::Time start();
+
+  absl::Time finish();
+
+  double micro_seconds();
+
+  int64 done();
+
+  int64 bytes();
+
+ private:
+  absl::Time start_;
+  absl::Time finish_;
+  double micro_seconds_;
+  int64 done_;
+  int64 next_report_;
+  int64 bytes_;
```
add docstring for these members and accessors.

BrianSong replied:
Sure! Fixed!

hughmiao commented on ml_metadata/tools/mlmd_bench/stats.h:
```diff
@@ -20,13 +20,55 @@ limitations under the License.
   int64 transferred_bytes;
 };
 
+class Stats {
```
nit, consider to call it ThreadStats, probably is more clear. Please add some class docstring to describe its usage.

BrianSong replied:
Sounds good! Fixed!

hughmiao commented on ml_metadata/tools/mlmd_bench/stats.cc:
```diff
@@ -0,0 +1,95 @@
+  bytes_ += op_stats.transferred_bytes;
+  micro_seconds_ += op_stats.elapsed_time / absl::Microseconds(1);
+  done_++;
+  // Reports the current progress.
+  if (total_done >= next_report_) {
+    if (next_report_ < 1000) {
+      next_report_ += 100;
+    } else if (next_report_ < 5000) {
+      next_report_ += 500;
+    } else if (next_report_ < 10000) {
+      next_report_ += 1000;
+    } else if (next_report_ < 50000) {
+      next_report_ += 5000;
+    } else if (next_report_ < 100000) {
+      next_report_ += 10000;
+    } else if (next_report_ < 500000) {
+      next_report_ += 50000;
+    } else {
+      next_report_ += 100000;
+    }
```
for these report boundary,  simply use a set<int> x of {1000, 5000, 10000, ...}
use the += upper_bound(next_report_) / 10 ? if the rule is not always / 10, use a map do keep k - v, then += v .

BrianSong replied:
Noted! Fixed!

hughmiao commented on ml_metadata/tools/mlmd_bench/stats.cc:
```diff
@@ -0,0 +1,95 @@
+  micro_seconds_ += other.micro_seconds_;
+  // Chooses the earliest start time and latest end time of each merged
+  // thread stats.
+  if (other.start_ < start_) start_ = other.start_;
+  if (other.finish_ > finish_) finish_ = other.finish_;
```
for these, simply use min or max , which reads better too

BrianSong replied:
Noted! Fixed!

hughmiao commented on ml_metadata/tools/mlmd_bench/thread_runner.h:
```diff
@@ -0,0 +1,45 @@
+  // failed.
+  tensorflow::Status Run(Benchmark& benchmark);
+
+ private:
+  int64 num_threads_;
+  ConnectionConfig mlmd_config_;
+  absl::Mutex lock_;
```
ditto, please add docstring, docstring, especially for the mutex member. 

often class mutex member uses the absl var/function annotations  to indicate which/how variables and functions related to this mutex [1][2]. 

from the cc, it looks the lock_ is not used across the class, consider to use a function local lock instead. 

[1] https://github.com/abseil/abseil-cpp/blob/master/absl/base/thread_annotations.h
[2] https://abseil.io/docs/cpp/guides/synchronization

BrianSong replied:
Yes, maybe we can discuss how to achieve this without the mutex.

hughmiao commented on ml_metadata/tools/mlmd_bench/thread_runner.cc:
```diff
@@ -0,0 +1,114 @@
+                                          "mlmd_bench", num_threads_);
+      // `total_done` is used for reporting progress along the way.
+      int64 total_done = 0;
+      int64 thread_index = 0;
+      for (int64 i = 0; i < num_threads_; ++i) {
+        pool.Schedule([&]() {
+          // Uses a lock to ensure the consistence of concurrent reading and
+          // writing of `thread`.
+          lock_.Lock();
+          int64 curr_thread_index = thread_index;
```
instead of using a thread_index, and lock/unlock to increase it, you can simply pass the thread_stat_list[i] to the Schedule lambda, each thread will then use its own stats. also instead of a default capture, use explicit capture of the variable needed in the lambda, see style guide [1]

[1] https://google.github.io/styleguide/cppguide.html#Lambda_expressions

BrianSong replied:
Yeah, this is better! Fixed! Thanks :)

hughmiao commented on ml_metadata/tools/mlmd_bench/thread_runner.cc:
```diff
@@ -0,0 +1,114 @@
+              continue;
+            }
+            work_items_index++;
+            lock_.Lock();
+            total_done++;
+            lock_.Unlock();
```
qq, this looks a barrier will affect all threads and serialize all functions, what're other alternatives to do the cmd reporting?

BrianSong replied:
Yes, we can discuss about it.

hughmiao reviewed (changes requested):
thanks, Brian. left some first round comments apart from the tests.

BrianSong committed:
Refactors benchmark + stats + thread runner according to reviews.

hughmiao commented on ml_metadata/tools/mlmd_bench/benchmark.h:
```diff
@@ -0,0 +1,48 @@
+ private:
+  // A list of executable workloads.
+  std::vector<std::unique_ptr<WorkloadBase>> workloads_;
+  // A list of number of operations for the executable workloads.
+  std::vector<int64> workloads_num_operations_;
```
drop?

hughmiao replied:
not done?  consider to push the new changes, will make another pass before approval

hughmiao commented on ml_metadata/tools/mlmd_bench/benchmark.h:
```diff
@@ -0,0 +1,48 @@
+
+  WorkloadBase* workload(int64 workload_index);
+
+  int64 workload_num_operations(int64 workload_index);
```
drop?

hughmiao commented on ml_metadata/tools/mlmd_bench/benchmark.h:
```diff
@@ -0,0 +1,48 @@
+
+  int64 workload_num_operations(int64 workload_index);
+
+  int64 workloads_size();
```
nit, num_workloads()

hughmiao commented on ml_metadata/tools/mlmd_bench/benchmark.cc:
```diff
@@ -0,0 +1,62 @@
+    std::unique_ptr<FillTypes> fill_types(new FillTypes(
+        workload_config.fill_types_config(), workload_config.num_operations()));
+    workloads.push_back(std::move(fill_types));
+    workloads_num_operations.push_back(workload_config.num_operations());
```
drop

hughmiao commented on ml_metadata/tools/mlmd_bench/benchmark.cc:
```diff
@@ -0,0 +1,62 @@
+                    std::vector<std::unique_ptr<WorkloadBase>>& workloads,
+                    std::vector<int64>& workloads_num_operations) {
+  if (workload_config.has_fill_types_config()) {
+    std::unique_ptr<FillTypes> fill_types(new FillTypes(
```
absl::make_unique [1]

[1] https://abseil.io/tips/126

hughmiao commented on ml_metadata/tools/mlmd_bench/benchmark.cc:
```diff
@@ -0,0 +1,62 @@
+  // workload.
+  for (const WorkloadConfig& workload_config :
+       mlmd_bench_config.workload_configs()) {
+    CreateWorkload(workload_config, workloads_, workloads_num_operations_);
```
workloads_.reserve(mlmd_bench_config.workload_configs_size()) 

for (int i = 0; ...; ..) {
  CreateWorkload(workload_config, workloads[i]);
}
// ----
CreateWorkload(workload, std::unique_ptr<WorkloadBase>& workload) {
    workload = make_unique ...
}

BrianSong replied:
Yes, I change the .reserve() to .resize() to avoid Segmentation fault when calling workloads_[i].

hughmiao commented on ml_metadata/tools/mlmd_bench/benchmark_test.cc:
```diff
@@ -0,0 +1,45 @@
+  Benchmark benchmark(mlmd_bench_config);
+  // Checks that indeed one workload has been created according to the
+  // `workload_configs`.
+  EXPECT_STREQ(benchmark.workload(0)->GetName().c_str(), "fill_artifact_type");
```
EXPECT_EQ(benchmark.num_workloads(), mlmd_bench_config..._size())

hughmiao commented on ml_metadata/tools/mlmd_bench/benchmark_test.cc:
```diff
@@ -0,0 +1,45 @@
+  // `workload_configs`.
+  EXPECT_STREQ(benchmark.workload(0)->GetName().c_str(), "fill_artifact_type");
+}
+
```
// constructor .. 
EXPECT_DEATH( ... ) ... // oss gtest

BrianSong replied:
Yes, for this issue, it seems that the wrong workload cannot even pass the testing::ParseTextProtoOrDie<>(), not to even mention enter into the benchmark. Should we still test it? Wdyt?

hughmiao replied:
i think it is optional. instead of parsing text, you can use proto constructor to try it. let's omit it from the cl.

hughmiao commented on ml_metadata/tools/mlmd_bench/stats.h:
```diff
@@ -20,13 +20,71 @@ limitations under the License.
+  // Records the finish time of current thread stats.
+  absl::Time finish_;
+  // Records the elapsed microseconds of current thread stats.
+  double micro_seconds_;
```
absl::Duration

hughmiao commented on ml_metadata/tools/mlmd_bench/stats.cc:
```diff
@@ -0,0 +1,90 @@
+  micro_seconds_ += op_stats.elapsed_time / absl::Microseconds(1);
+  done_++;
+  // Reports the current progress.
+  std::vector<int> report_thresholds{1000,   5000,   10000,  50000,
```
consider to use a static array, see [1] for common patterns of static var
int[] {} 

[1] https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables

hughmiao commented on ml_metadata/tools/mlmd_bench/stats.cc:
```diff
@@ -0,0 +1,90 @@
+  std::vector<int> report_thresholds{1000,   5000,   10000,  50000,
+                                     100000, 500000, 1000000};
+  int threshold_index = 0;
+  if (total_done >= next_report_) {
```
if (total_done < next_report_) {
    return;
}

next_report_ += ..
...

}

hughmiao commented on ml_metadata/tools/mlmd_bench/stats.cc:
```diff
@@ -0,0 +1,90 @@
+  micro_seconds_ += other.micro_seconds_;
+  // Chooses the earliest start time and latest end time of each merged
+  // thread stats.
+  start_ = std::min(start_, other.start_);
```
other.start()

hughmiao commented on ml_metadata/tools/mlmd_bench/stats.cc:
```diff
@@ -0,0 +1,90 @@
+    // earliest start time of each thread) instead of the sum of per-thread
+    // elapsed times.
+    double elapsed_seconds = micro_seconds_ * 1e-6;
+    char rate[100];
```
absl::StrCat ..
asbl::StrFormat ..

hughmiao commented on ml_metadata/tools/mlmd_bench/stats_test.cc:
```diff
@@ -0,0 +1,124 @@
+  stats2.Stop();
+  absl::SleepFor(sleep_time);
+  stats3.Stop();
+  absl::SleepFor(sleep_time);
```
drop

hughmiao commented on ml_metadata/tools/mlmd_bench/thread_runner.h:
```diff
@@ -0,0 +1,45 @@
+  // Number of threads for the thread runner.
+  int64 num_threads_;
+  // Connection configuration that will be used to create the MetadataStore.
+  ConnectionConfig mlmd_config_;
```
const

hughmiao commented on ml_metadata/tools/mlmd_bench/thread_runner.h:
```diff
@@ -0,0 +1,45 @@
+
+ private:
+  // Number of threads for the thread runner.
+  int64 num_threads_;
```
const

hughmiao commented on ml_metadata/tools/mlmd_bench/thread_runner.cc:
```diff
@@ -0,0 +1,107 @@
+      mlmd_config_(mlmd_bench_config.mlmd_config()) {}
+
+// Execution unit of mlmd_bench. Returns detailed error if query executions
+// failed.
```
drop the public comments

hughmiao commented on ml_metadata/tools/mlmd_bench/thread_runner.cc:
```diff
@@ -0,0 +1,107 @@
+tensorflow::Status ThreadRunner::Run(Benchmark& benchmark) {
+  int64 workloads_size = benchmark.workloads_size();
+  for (int workload_index = 0; workload_index < workloads_size;
+       ++workload_index) {
```
format a bit

hughmiao commented on ml_metadata/tools/mlmd_bench/thread_runner.cc:
```diff
@@ -0,0 +1,107 @@
+        benchmark.workload_num_operations(workload_index) / num_threads_;
+    std::unique_ptr<MetadataStore> set_up_store;
+    TF_RETURN_IF_ERROR(CreateMetadataStore(mlmd_config_, &set_up_store));
+    TF_RETURN_IF_ERROR(workload->SetUp(set_up_store.get()));
```
move up

hughmiao commented on ml_metadata/tools/mlmd_bench/thread_runner.cc:
```diff
@@ -0,0 +1,107 @@
+      tensorflow::thread::ThreadPool pool(tensorflow::Env::Default(),
+                                          "mlmd_bench", num_threads_);
+      // `total_done` is used for reporting progress along the way.
+      int64 total_done = 0;
```
approx_total_done

hughmiao commented on ml_metadata/tools/mlmd_bench/thread_runner.cc:
```diff
@@ -0,0 +1,107 @@
+      // `total_done` is used for reporting progress along the way.
+      int64 total_done = 0;
+      for (int64 i = 0; i < num_threads_; ++i) {
+        pool.Schedule([op_per_thread, workload, mlmd_config = mlmd_config_, i,
```
passing in this, mlmd_config = mlmd_config_

hughmiao commented on ml_metadata/tools/mlmd_bench/thread_runner.cc:
```diff
@@ -0,0 +1,107 @@
+      int64 total_done = 0;
+      for (int64 i = 0; i < num_threads_; ++i) {
+        pool.Schedule([op_per_thread, workload, mlmd_config = mlmd_config_, i,
+                       &thread_stats_list, &total_done]() {
```
&thread_stats_list[i], start_index,

BrianSong replied:
Yes, I found that passing &thread_stats_list[i] will raise compiler to complain. So I take the reference before the pool schedule, store it inside a curr_thread_stats and pass that into the lambda.

hughmiao commented on ml_metadata/tools/mlmd_bench/thread_runner.cc:
```diff
@@ -0,0 +1,107 @@
+          // Each thread uses a different MLMD client instance to talk to
+          // the same back-end.
+          std::unique_ptr<MetadataStore> store;
+          CreateMetadataStore(mlmd_config, &store);
```
TF_CHECK_OK(...) / TF_RETURN_IF_ERROR
or 
.IgnoreError()

BrianSong replied:
Hi Hui, a qq. I found that CreateMetadataStore() may also return Abort error when running in multi-thread, so I add some additional logic to ensure the store is created. PTAL.

hughmiao commented on ml_metadata/tools/mlmd_bench/thread_runner.cc:
```diff
@@ -0,0 +1,107 @@
+          // the same back-end.
+          std::unique_ptr<MetadataStore> store;
+          CreateMetadataStore(mlmd_config, &store);
+          int64 start_index = op_per_thread * i;
```
drop

hughmiao commented on ml_metadata/tools/mlmd_bench/thread_runner.cc:
```diff
@@ -0,0 +1,107 @@
+            tensorflow::Status status =
+                workload->RunOp(work_items_index, store.get(), op_stats);
+            // Handles abort issues for concurrent writing to the db.
+            if (!status.ok()) {
```
check Aborted error code , 

TF_CHECK_OK / TF_RETURN_IF_ERROR if it is not Aborted ..

hughmiao commented on ml_metadata/tools/mlmd_bench/thread_runner.cc:
```diff
@@ -0,0 +1,107 @@
+    for (int64 i = 1; i < num_threads_; ++i) {
+      thread_stats_list[0].Merge(thread_stats_list[i]);
+    }
+    // Reports the metrics of interests.
```
// TODO(briansong) return the report as a proto

hughmiao commented on ml_metadata/tools/mlmd_bench/thread_runner.h:
```diff
@@ -0,0 +1,45 @@
+// It takes the benchmark and runs the workloads.
+class ThreadRunner {
+ public:
+  ThreadRunner(const MLMDBenchConfig& mlmd_bench_config);
```
(const ConnectionConfig& , int64 num_thread)

hughmiao commented on ml_metadata/tools/mlmd_bench/thread_runner.cc:
```diff
@@ -0,0 +1,107 @@
+// After the each thread has finished the execution, the workload stats will be
+// generated by merging all the thread stats inside the `thread_stats_list`. The
+// performance of the workload will be reported according to the workload stats.
+tensorflow::Status ThreadRunner::Run(Benchmark& benchmark) {
```
add the output parameter Summary& summary

BrianSong replied:
Added TODO. Plan to add the Summary in the next PR along with the binary mlmd_bench.cc where the summary will be written into a file in disk. Wdyt?

hughmiao replied:
sgtm

hughmiao commented on ml_metadata/tools/mlmd_bench/thread_runner_test.cc:
```diff
@@ -0,0 +1,102 @@
+
+namespace ml_metadata {
+namespace {
+
```
move everything to an anonymous function

TestThreadRunner(num_thread) ...

hughmiao reviewed (changes requested):
thanks, Brian.

hughmiao commented on ml_metadata/tools/mlmd_bench/BUILD:
```diff
@@ -75,7 +150,7 @@ cc_library(
     deps = [
         "@com_google_absl//absl/types:variant",
-        "//ml_metadata/metadata_store",
+        "//ml_metadata/metadata_store:metadata_store",
```
Brian, the previous CL update node just committed to the github master [1], please sync.  there's some file changes which may conflict with some of the BUILD

[1] https://github.com/google/ml-metadata/commit/1a0e05e9431f00023bb691ef9b85d51c70c0ab55

BrianSong replied:
Sure! Already synced and fixed the conflict. Thanks, Hui!

BrianSong committed:
Refactors thread runner + benchmark + stats according to 2nd round review.

BrianSong committed:
Sync with head.

hughmiao commented on ml_metadata/tools/mlmd_bench/benchmark.cc:
```diff
@@ -0,0 +1,58 @@
+// Creates the executable workload given `workload_config`.
+void CreateWorkload(const WorkloadConfig& workload_config,
+                    std::unique_ptr<WorkloadBase>& workload) {
+  if (workload_config.has_fill_types_config()) {
+    workload = absl::make_unique<FillTypes>(FillTypes(
+        workload_config.fill_types_config(), workload_config.num_operations()));
+  } else {
+    LOG(FATAL) << "Cannot find corresponding workload!";
+  }
+}
```
nit, 
if (! ...) { LOG(FATAL) << ... }
workload = ..

hughmiao commented on ml_metadata/tools/mlmd_bench/benchmark.h:
```diff
@@ -0,0 +1,46 @@
+  WorkloadBase* workload(int64 workload_index);
+
+  // Returns the number of executable workloads existed inside benchmark.
+  const int64 num_workloads();
```
you probably mean

int64 num_workloads() const; ?

BrianSong replied:
Yes, my bad. Fixed!

hughmiao commented on ml_metadata/tools/mlmd_bench/thread_runner.cc:
```diff
@@ -0,0 +1,120 @@
+            // If the error is not Abort error, break the current process.
+            if (!store_status.ok() &&
+                store_status.code() != tensorflow::error::ABORTED) {
+              TF_CHECK_OK(store_status);
```
TF_RETURN_IF_ERROR

hughmiao commented on ml_metadata/tools/mlmd_bench/thread_runner.cc:
```diff
@@ -0,0 +1,120 @@
+                workload->RunOp(work_items_index, store.get(), op_stats);
+            // If the error is not Abort error, break the current process.
+            if (!status.ok() && status.code() != tensorflow::error::ABORTED) {
+              TF_CHECK_OK(status);
```
ditto

BrianSong replied:
Yes, I tried to change the TF_CHECK_OK() to TF_RETURN_IF_ERROR() inside the pool Schedule(), there will always be a Segmentation fault. I also try to print out the error message, but it seems that the program does not even come into the if statement. Wdyt? Does the error happen because we cannot do  TF_RETURN_IF_ERROR()  in a pool lambda function?

hughmiao commented on ml_metadata/tools/mlmd_bench/thread_runner.cc:
```diff
@@ -0,0 +1,120 @@
+          tensorflow::Status store_status =
+              CreateMetadataStore(mlmd_config_, &store);
+          // Handles abort issues for concurrent creating `store` of the db.
+          while (!store_status.ok()) {
+            store_status = CreateMetadataStore(mlmd_config_, &store);
+            // If the error is not Abort error, break the current process.
+            if (!store_status.ok() &&
+                store_status.code() != tensorflow::error::ABORTED) {
+              TF_CHECK_OK(store_status);
+            }
+          }
```
create the store in before the pool (i.e., after 51~53) and pass store[i] to the lambda. in that way, these store are opened sequentially.

BrianSong replied:
Good idea! Fixed!

hughmiao commented on ml_metadata/tools/mlmd_bench/thread_runner.cc:
```diff
@@ -0,0 +1,120 @@
+    thread_stats_list[0].Report(workload->GetName());
+  }
+  return tensorflow::Status::OK();
+}
```
consider to move some part to internal functions, currently this method is very long.

BrianSong replied:
My bad! Fixed! Will be every carefully about this in the future.

hughmiao reviewed (changes requested):
overall lgtm, some comments in the thread runner.

BrianSong committed:
Revise the thread_runner.cc according to reviews.

hughmiao commented on ml_metadata/tools/mlmd_bench/stats.cc:
```diff
@@ -0,0 +1,85 @@
+
+void ThreadStats::Start() { start_ = absl::Now(); }
+
+void ThreadStats::Update(const OpStats& op_stats, int64& approx_total_done) {
```
pass by value for primitive types. instead of int64&, use int64

hughmiao commented on ml_metadata/tools/mlmd_bench/stats.h:
```diff
@@ -20,13 +20,73 @@ limitations under the License.
+  void Start();
+
+  // Updates the current thread stats with op_stats.
+  void Update(const OpStats& op_stats, int64& approx_total_done);
```
ditto

hughmiao committed:
Update thread_runner.cc

some nit edits on names, and passing pointers instead of unique_ptr<>&. please run test before submitting.

hughmiao reviewed (approved):
please fix the nit comments and run tests before merging.

BrianSong committed:
Finalizes 3rd PR: adding benchmark + thread runner + stats.
